src/models/tree/build_tree.cu

```diff
@@ -135,56 +135,64 @@ __global__ static void __launch_bounds__(THREADS_PER_BLOCK, MIN_CTAS_PER_SM)
+                      legate::AccessorRO<double, 3> h,
+                      size_t n_outputs,
+                      SparseSplitProposals<TYPE> split_proposals,
+                      NodeBatch batch,
+                      Histogram<IntegerGPair> histogram,
+                      legate::Buffer<IntegerGPair, 2> node_sums,
+                      GradientQuantiser quantiser)
 {
-  // block dimensions are (THREADS_PER_BLOCK, 1, 1)
-  // each thread processes ELEMENTS_PER_THREAD samples and FEATURES_PER_BLOCK features
-  // the features to process are defined via blockIdx.y
-
-  // further improvements:
-  // * quantize values to work with int instead of double
-
-#pragma unroll
-  for (int32_t elementIdx = 0; elementIdx < ELEMENTS_PER_THREAD; ++elementIdx) {
-    // within each iteration a (THREADS_PER_BLOCK, FEATURES_PER_BLOCK)-block of
-    // data from X is processed.
-
-    // check if thread has actual work to do
-    int64_t idx      = (blockIdx.x + elementIdx * gridDim.x) * THREADS_PER_BLOCK + threadIdx.x;
-    bool validThread = idx < batch.InstancesInBatch();
-    if (!validThread) continue;
-    auto [sampleNode, localSampleId] = batch.instances_begin[idx];
-    int64_t globalSampleId           = localSampleId + sample_offset;
-
-    bool computeHistogram = ComputeHistogramBin(
-      sampleNode, node_sums, histogram.ContainsNode(BinaryTree::Parent(sampleNode)));
-
-    for (int32_t output = 0; output < n_outputs; output++) {
-      auto gpair_quantised =
-        quantiser.Quantise({g[{globalSampleId, 0, output}], h[{globalSampleId, 0, output}]});
-      for (int32_t featureIdx = 0; featureIdx < FEATURES_PER_BLOCK; featureIdx++) {
-        int32_t feature = featureIdx + blockIdx.y * FEATURES_PER_BLOCK;
-        if (computeHistogram && feature < n_features) {
-          auto x_value = X[{globalSampleId, feature, 0}];
-          auto bin_idx = split_proposals.FindBin(x_value, feature);
-
-          // bin_idx is the first sample that is larger than x_value
+  constexpr int32_t WarpSize = 32;
+  const int32_t warp_id      = threadIdx.x / WarpSize;
+  const int32_t lane_id      = threadIdx.x % WarpSize;
+
+  const int32_t localIdx = blockIdx.x * TPB + warp_id * WarpSize + lane_id;
+
+  // prefetch sampleNode information for all 32 ids
+  auto [sampleNode_lane, localSampleId_lane] = (localIdx < batch.InstancesInBatch())
+                                                 ? batch.instances_begin[localIdx]
+                                                 : cuda::std::make_tuple(-1, -1);
+  const bool computeHistogram =
+    localIdx < batch.InstancesInBatch() &&
+    ComputeHistogramBin(
+      sampleNode_lane, node_sums, histogram.ContainsNode(BinaryTree::Parent(sampleNode_lane)));
+
+  // mask contains all sample bits of the next 32 ids that need to be bin'ed
+  auto lane_mask = ballot(computeHistogram);
+
+  // reverse to use __clz instead of __ffs
+  lane_mask = __brev(lane_mask);
+
+  while (lane_mask) {
```

Comment:
One way of getting the bit fidding and intrinsics abstracted out:

```
template<typname T>
class WarpQueue{
private:
T x;
int32_t mask;
public:
WarpQueue(T x, bool active) ...
bool Empty(){}
T Pop(){}
}

WarpQueue warp_items(local_items, active);
while(!warp_items.empty()){
  auto [node, sampleId] = warp_items.Pop();
}

```

Feel free to ignore.

Reply:
The suggestion is only for the offsets.